### PR TITLE
Add play media action

### PR DESCRIPTION
**Pull request by bramkragten**
## Proposed change

Add play media action + media selector

~Needs: https://github.com/home-assistant/core/pull/66663~

Needs: https://github.com/home-assistant/core/pull/66672

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/components/ha-selector/ha-selector-media.ts [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+import {
+  MediaClassBrowserSettings,
+  MediaPickedEvent,
+  SUPPORT_BROWSE_MEDIA,
+} from "../../data/media-player";
+import { MediaSelector } from "../../data/selector";
+import { HomeAssistant } from "../../types";
+import "../ha-alert";
+import "../ha-form/ha-form";
+import { HaFormSchema } from "../ha-form/types";
```
Comment (zsarnett) [inline src/components/ha-selector/ha-selector-media.ts:16]:
```suggestion
import type { MediaSelector } from "../../data/selector";
import type { HomeAssistant } from "../../types";
import "../ha-alert";
import "../ha-form/ha-form";
import type { HaFormSchema } from "../ha-form/types";
```

File: src/components/ha-selector/ha-selector-media.ts [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+                this._manual ? "browse_media" : "manual"
+              }`
+            )}
+          </mwc-button>`
+        : ""}`;
+  }
+
+  private _computeLabelCallback(schema: HaFormSchema): string {
+    return this.hass.localize(`ui.components.selectors.media.${schema.name}`);
+  }
```
Comment (zsarnett) [inline src/components/ha-selector/ha-selector-media.ts:149]:
```suggestion
  private _computeLabelCallback = (schema: HaFormSchema): string => {
    return this.hass.localize(`ui.components.selectors.media.${schema.name}`);
  }
```

File: src/panels/config/automation/action/types/ha-automation-action-media_content_id.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      <ha-selector-media
+        .hass=${this.hass}
+        .value=${this.action}
+      ></ha-selector-media>
+    `;
+  }
+
+  static get styles(): CSSResultGroup {
+    return css``;
+  }
```
Comment (zsarnett) [inline src/panels/config/automation/action/types/ha-automation-action-media_content_id.ts:32]:
```suggestion
```

File: src/translations/en.json [hunk 4]
```diff
@@ -317,6 +317,17 @@
       "copied_clipboard": "Copied to clipboard"
     },
     "components": {
+      "selectors": {
+        "media": {
+          "pick_media_player": "Select media player",
+          "browse_not_supported": "This media player does not support browsing media, but you can enter the media ID and type manually.",
+          "pick_media": "Pick media",
+          "browse_media": "Browse media",
+          "manual": "Enter media ID manual",
```
Comment (zsarnett) [inline src/translations/en.json:326]:
```suggestion
          "manual": "Manually enter Media ID",
```

File: src/components/ha-selector/ha-selector-media.ts (hunk 2 above)
Comment (bramkragten) [reply src/components/ha-selector/ha-selector-media.ts:149]: why?

Comment (zsarnett) [reply src/components/ha-selector/ha-selector-media.ts]: https://github.com/home-assistant/frontend/pull/11648#discussion_r804090724

File: src/components/ha-selector/ha-selector-media.ts [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+  @state() private _manual = false;
+
+  @state() private _thumbnailUrl?: string | null;
+
+  willUpdate(changedProps) {
+    if (
+      changedProps.has("value") &&
+      this.value?.extra?.thumbnail !==
+        changedProps.get("value")?.extra?.thumbnail
+    ) {
```
Comment (balloob) [inline src/components/ha-selector/ha-selector-media.ts:46]: There are a lot of `?` in this method making it hard to follow. Can you extract the extraThumbnail and extraThumbnailOld to variables?

File: src/components/ha-selector/ha-selector-media.ts [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+        .hass=${this.hass}
+        .value=${this.value?.entity_id}
+        .label=${this.label ||
+        this.hass.localize("ui.components.selectors.media.pick_media_player")}
+        .disabled=${this.disabled}
+        include-domains='["media_player"]'
+        allow-custom-entity
+        @value-changed=${this._entityChanged}
+      ></ha-entity-picker>
+      ${this._manual || (stateObj && !supportsBrowse)
```
Comment (balloob) [inline src/components/ha-selector/ha-selector-media.ts:82]: Why check if stateObj is defined?

Comment (bramkragten) [reply src/components/ha-selector/ha-selector-media.ts:82]: If we dont have a state, we cant check if it supports browse, so assume it does

**Review by zsarnett — DISMISSED** [review]

File: src/components/ha-selector/ha-selector-media.ts (hunk 6 above)
Comment (balloob) [reply src/components/ha-selector/ha-selector-media.ts:79]: You already guard for that when defining `supportsBrowse`

Comment (bramkragten) [reply src/components/ha-selector/ha-selector-media.ts:79]: yes, but if there is a state, and we checked if it supported browsing, and it didnt, then we get here.

If `supportsBrowse` is false because we didnt have a `stateObj`, we render the media browse UI

Comment (bramkragten) [reply src/components/ha-selector/ha-selector-media.ts:79]: Changed it be clearer

File: gallery/src/pages/components/ha-selector.ts [hunk 7]
```diff
@@ -73,6 +73,7 @@ const SCHEMAS: {
         selector: { select: { options: ["Option 1", "Option 2"] } },
       },
       icon: { name: "Icon", selector: { icon: {} } },
+      media: { name: "Media", selector: { media: {} } },
```
Comment (balloob) [inline gallery/src/pages/components/ha-selector.ts:76]: Can we also add it to the ha-form gallery page. It has a form with all selectors.

File: src/components/ha-selector/ha-selector-media.ts [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+      !this.value?.entity_id ||
+      (stateObj && supportsFeature(stateObj, SUPPORT_BROWSE_MEDIA));
+
+    return html`<ha-entity-picker
+        .hass=${this.hass}
+        .value=${this.value?.entity_id}
+        .label=${this.label ||
+        this.hass.localize("ui.components.selectors.media.pick_media_player")}
+        .disabled=${this.disabled}
+        include-domains='["media_player"]'
```
Comment (balloob) [inline src/components/ha-selector/ha-selector-media.ts:76]: should `ha-entity-picker` support entity selectors in a future PR?

File: gallery/src/pages/components/ha-selector.ts (hunk 7 above)
Comment (balloob) [reply gallery/src/pages/components/ha-selector.ts:76]: Also, we have automation/selectors.ts too. We have 2 design pages for selectors? 🤦 

https://620d65c9f5140e8acf3cf643--home-assistant-gallery.netlify.app/#components/ha-selector

Changing between browse and manual logs a JS error.

File: src/components/ha-selector/ha-selector-media.ts [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+                  `}
+            </div>
+            <div class="title">
+              ${!this.value?.media_content_id
+                ? this.hass.localize("ui.components.selectors.media.pick_media")
+                : this.value.metadata?.title || this.value.media_content_id}
+            </div>
+          </ha-card>`}
+      ${supportsBrowse
+        ? html`<mwc-button @click=${this._switchManual}
```
Comment (balloob) [inline src/components/ha-selector/ha-selector-media.ts:150]: I don't like this button. I think that if we want to offer this, it should be part of the "pick media" dialog instead.

Comment (balloob) [reply src/components/ha-selector/ha-selector-media.ts:150]: We could make a special navigation ID called `manual://` that will show this form

Comment (bramkragten) [reply src/components/ha-selector/ha-selector-media.ts:150]: You also want to do that for entities not supporting media browser?

File: src/components/ha-selector/ha-selector-media.ts (hunk 8 above)
Comment (bramkragten) [reply src/components/ha-selector/ha-selector-media.ts:76]: We have `ha-entity-selector` for that?

File: src/components/media-player/show-media-browser-dialog.ts [hunk 10]
```diff
@@ -3,13 +3,15 @@ import {
...
   MediaPlayerBrowseAction,
 } from "../../data/media-player";
+import { MediaPlayerItemId } from "./ha-media-player-browse";
 
 export interface MediaPlayerBrowseDialogParams {
   action: MediaPlayerBrowseAction;
   entityId: string;
   mediaPickedCallback: (pickedMedia: MediaPickedEvent) => void;
   mediaContentId?: string;
   mediaContentType?: string;
```
Comment (balloob) [inline src/components/media-player/show-media-browser-dialog.ts:13]: I think that we can remove this, they are not used.

File: src/data/media-player.ts [hunk 11]
```diff
@@ -147,6 +148,7 @@ export const MediaClassBrowserSettings: {
 
 export interface MediaPickedEvent {
   item: MediaPlayerItem;
+  navigateIds?: MediaPlayerItemId[];
```
Comment (balloob) [inline src/data/media-player.ts:151]: should we make this mandatory and instead have the TTS form use a different event?

File: src/data/script.ts [hunk 12]
```diff
@@ -246,6 +255,12 @@ export const getActionType = (action: Action): ActionType => {
       ) {
         return "activate_scene";
       }
+      if (
+        (action as PlayMediaAction).service === "media_player.play_media" &&
+        !Array.isArray((action as PlayMediaAction)?.target?.entity_id)
```
Comment (balloob) [inline src/data/script.ts:260]: I think that we should make sure that the number of allowed keys only includes the ones we support.

File: src/components/ha-selector/ha-selector-media.ts (hunk 8 above)
Comment (balloob) [reply src/components/ha-selector/ha-selector-media.ts:76]: true true.

File: src/components/ha-selector/ha-selector-media.ts (hunk 9 above)
Comment (balloob) [reply src/components/ha-selector/ha-selector-media.ts:150]: yeah

Comment (balloob) [reply src/components/ha-selector/ha-selector-media.ts:150]: We can do that in a future PR if you want.

File: src/translations/en.json [hunk 13]
```diff
@@ -317,6 +317,17 @@
       "copied_clipboard": "Copied to clipboard"
     },
     "components": {
+      "selectors": {
+        "media": {
+          "pick_media_player": "Select media player",
+          "browse_not_supported": "This media player does not support browsing media, but you can enter the media ID and type manually.",
```
Comment (balloob) [inline src/translations/en.json:323]:
```suggestion
          "browse_not_supported": "Media player does not support browsing media.",
```

**Review by balloob — DISMISSED** [review]

**Review by balloob — APPROVED** [review]
